client/client-lib/src/api.rs

```diff
@@ -1,171 +1,575 @@
+use std::pin::Pin;
 use std::sync::Arc;
 use std::time::Duration;
+use std::{cmp, ops, result};
 
 use async_trait::async_trait;
-use bitcoin::{Address, Amount};
-use bitcoin_hashes::sha256::Hash as Sha256Hash;
 use fedimint_api::config::ClientConfig;
-use fedimint_api::core::{
-    LEGACY_HARDCODED_INSTANCE_ID_LN, LEGACY_HARDCODED_INSTANCE_ID_MINT,
-    LEGACY_HARDCODED_INSTANCE_ID_WALLET,
-};
 use fedimint_api::module::registry::ModuleDecoderRegistry;
-use fedimint_api::task::{sleep, RwLock, RwLockWriteGuard};
-use fedimint_api::{dyn_newtype_define, NumPeers, OutPoint, PeerId, TransactionId};
+use fedimint_api::task::{RwLock, RwLockWriteGuard};
+use fedimint_api::{dyn_newtype_define, NumPeers, PeerId, TransactionId};
 use fedimint_core::epoch::{SerdeEpochHistory, SignedEpochOutcome};
-use fedimint_core::modules::ln::contracts::incoming::IncomingContractOffer;
-use fedimint_core::modules::ln::contracts::ContractId;
-use fedimint_core::modules::ln::{ContractAccount, LightningGateway};
-use fedimint_core::modules::mint::db::ECashUserBackupSnapshot;
-use fedimint_core::modules::wallet::PegOutFees;
-use fedimint_core::outcome::legacy::{OutputOutcome, TryIntoOutcome};
 use fedimint_core::outcome::TransactionStatus;
-use fedimint_core::transaction::legacy::Transaction as LegacyTransaction;
 use fedimint_core::transaction::SerdeTransaction;
-use fedimint_core::CoreError;
 use futures::stream::FuturesUnordered;
-use futures::StreamExt;
+use futures::{Future, StreamExt};
 #[cfg(not(target_family = "wasm"))]
 use jsonrpsee_core::client::CertificateStore;
 use jsonrpsee_core::client::ClientT;
 use jsonrpsee_core::Error as JsonRpcError;
-use jsonrpsee_types::error::CallError as RpcCallError;
 #[cfg(target_family = "wasm")]
 use jsonrpsee_wasm_client::{Client as WsClient, WasmClientBuilder as WsClientBuilder};
 #[cfg(not(target_family = "wasm"))]
 use jsonrpsee_ws_client::{WsClient, WsClientBuilder};
-use serde::{Deserialize, Serialize};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
 use thiserror::Error;
 use threshold_crypto::PublicKey;
-use tracing::{debug, error, instrument, trace, warn};
+use tokio::time::sleep;
+use tracing::{debug, error, instrument};
 use url::Url;
 
 use crate::module_decode_stubs;
-use crate::query::{
-    CurrentConsensus, EventuallyConsistent, QueryStep, QueryStrategy, Retry404, UnionResponses,
-    UnionResponsesSingle, ValidHistory,
-};
+use crate::query::{CurrentConsensus, QueryStep, QueryStrategy, UnionResponses, ValidHistory};
+use crate::LegacyTransaction;
 
 #[cfg_attr(target_family = "wasm", async_trait(? Send))]
 #[cfg_attr(not(target_family = "wasm"), async_trait)]
-pub trait IFederationApi: Debug + Send + Sync {
-    /// Fetch the outcome of an entire transaction
-    async fn fetch_tx_outcome(&self, tx: TransactionId) -> Result<TransactionStatus>;
+pub trait IFederationApi: Debug {
+    /// List of available federation members
+    fn members(&self) -> &BTreeSet<PeerId>;
```

Comment:
Unclear if this would be the ones we currently have a connection to or the ones we could open one to.

Reply:
Oh, I thought it's already all and the underlying impl is responsible for maintaining connectivity. Will add a comment.

Reply:
I mean it kinda is responsible, but not always able to, so this fn could mean either.